### PR TITLE
syncer: save empty SQL when online DDL is filtered

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
close https://github.com/pingcap/dm/issues/1656

### What is changed and how it works?
- save an empty SQL when online DDL is filtered, and remove it when read it from online DDL component
- use `?` not fmt.Sprintf to build a query

### Check List 

Tests 

 - Integration test

Code changes

 - Has persistent data change (not likely, because if there's quote in source-id, an error already occurred)

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: tests/online_ddl/conf/dm-task.yaml [hunk 1]
```diff
@@ -68,6 +70,14 @@ column-mappings:
     target-column: "id"
     arguments: ["2", "", "t"]
 
+filters:
+  filter-rule-index:
+    schema-pattern: "*"
+    table-pattern: "*"
+    events: ["create index","drop index"]
+    sql-pattern: ["ALTER\\s.*TABLE[\\s\\S]*ADD\\s+KEY", "ALTER\\s.*TABLE[\\s\\S]*DROP\\s+KEY"]
```
Comment (GMHDBJD) [inline tests/online_ddl/conf/dm-task.yaml:78]: How about filter index statement for special name, so we will have positive and negative test.

File: syncer/syncer.go [hunk 2]
```diff
@@ -1751,6 +1752,21 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
+			ec.tctx.L().Info("failed to parse a filtered SQL for online DDL", zap.String("SQL", originSQL))
+		}
+		for _, stmt := range stmts {
+			if _, ok := stmt.(ast.DDLNode); ok {
+				tableNames, err3 := parserpkg.FetchDDLTableNames(usedSchema, stmt)
+				if err3 != nil {
+					continue
+				}
+				// nolint:errcheck
+				s.onlineDDL.Apply(ec.tctx, tableNames, "", stmt)
```
Comment (GMHDBJD) [inline syncer/syncer.go:1767]: Should we check onlineDDL is nil?

File: syncer/syncer.go [hunk 3]
```diff
@@ -1751,6 +1752,21 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 		skipBinlogDurationHistogram.WithLabelValues("query", s.cfg.Name, s.cfg.SourceID).Observe(time.Since(ec.startTime).Seconds())
 		ec.tctx.L().Warn("skip event", zap.String("event", "query"), zap.String("statement", originSQL), zap.String("schema", usedSchema))
 		*ec.lastLocation = *ec.currentLocation // before record skip location, update lastLocation
+		// we try to insert an empty SQL to s.onlineDDL, because ignoring it will cause a "not found" error
+		stmts, err2 := parserpkg.Parse(parser2, originSQL, "", "")
+		if err2 != nil {
+			ec.tctx.L().Info("failed to parse a filtered SQL for online DDL", zap.String("SQL", originSQL))
+		}
+		for _, stmt := range stmts {
```
Comment (GMHDBJD) [inline syncer/syncer.go:1760]: if else?

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM. make build to fix fmt.

File: tests/online_ddl/conf/dm-task.yaml (hunk 1 above)
Comment (lance6716) [reply tests/online_ddl/conf/dm-task.yaml]: what's the "negative", do you mean filter a SQL that parser can't parse, and test `handle-error skip` it?

**Review by lichunzhu — COMMENTED** [review]
/lgtm

**Pull request merged**